test-fixture/src/sim/taildrop.rs

```diff
@@ -124,38 +213,92 @@ impl TailDrop {
     }
 
     /// Enqueue for sending.  Maybe.  If this overflows the queue, drop it instead.
+    /// If the queue is too deep, CE mark it before saving.
     fn maybe_enqueue(&mut self, d: Datagram, now: Instant) {
-        self.received += 1;
+        self.stats.received += 1;
         if self.next_deque.is_none() {
             // Nothing in the queue and nothing still sending.
             debug_assert!(self.queue.is_empty());
-            self.send(d, now);
+            self.send(d, now, 0);
         } else if self.used + self.size(&d) <= self.capacity {
+            self.stats.accumulate_usage(now, self.used);
             self.used += self.size(&d);
-            self.maxq = max(self.maxq, self.used);
+            self.stats.update_maxq(self.used);
+            let d = self.maybe_mark(d);
             self.queue.push_back(d);
         } else {
             qtrace!("taildrop dropping {} bytes", d.len());
-            self.dropped += 1;
+            self.stats.dropped += 1;
         }
     }
 
     /// If the last packet that was sending has been sent, start sending
     /// the next one.
     fn maybe_send(&mut self, now: Instant) {
-        if self.next_deque.as_ref().is_some_and(|t| *t <= now) {
+        if let Some((t, subns)) = &self.next_deque {
+            if now < *t {
+                return;
+            }
+
+            self.stats.accumulate_usage(now, self.used);
             if let Some(d) = self.queue.pop_front() {
                 self.used -= self.size(&d);
-                self.send(d, now);
+                self.send(d, *t, *subns);
             } else {
                 self.next_deque = None;
-                self.sub_ns_delay = 0;
             }
         }
     }
+
+    /// Apply ECN-CE markings to packets if they are ECT(0) marked.
+    /// This is classic, simple ECN using RED; we don't know about L4S yet.
+    fn maybe_mark(&mut self, dgram: Datagram) -> Datagram {
+        if self.ecn && Ecn::from(dgram.tos()) == Ecn::Ect0 && self.should_mark(self.used) {
+            qtrace!("taildrop marking {} bytes", dgram.len());
+            self.stats.marked += 1;
+            let tos = Tos::from((Dscp::from(dgram.tos()), Ecn::Ce));
+            Datagram::new(dgram.source(), dgram.destination(), tos, dgram.to_vec())
+        } else {
+            dgram
+        }
+    }
+
+    fn should_mark(&self, used: usize) -> bool {
+        // Apply RED which starts at 0 mark chance at 30% of the capacity.
+        // From there, follow a cubic curve that reaches 1 at 80% capacity.
+        // Cap at around 95% mark probability.
+        //
+        // let p = (2 * ((used / capacity) - 0.3));
+        // if rand(0, 1) < p.pow(3).clamp(0, 0.95) { mark(d) } else { d }
+        //
+        // This code scales that up by a factor of 1024 so we can use integers.
```

Comment:
Do we care that we can use integers? This isn't kernel code.

Reply:
The numbers we are using are big enough that they won't fit in the mantissa of `f64`.  That might not be a problem, because we don't care about the rounding error accumulation really.  It's more that the random number generator only produces an integer.  Teaching it to produce a float is harder than moving to ints here.  It's a little fiddly here, but not overly so.